src/raftstore/store/peer.rs

```diff
@@ -119,30 +122,55 @@ impl ReadIndexQueue {
     }
 
     /// update the read index of the requests that before the specified id.
-    fn advance(&mut self, id: &[u8], read_index: u64) {
-        if let Some(i) = self.reads.iter().position(|x| x.binary_id() == id) {
-            for pos in 0..=i {
-                let req = &mut self.reads[pos];
-                let index = req.read_index.get_or_insert(read_index);
-                if *index > read_index {
-                    *index = read_index;
+    fn advance_replica_reads(&mut self, states: &[ReadState]) {
+        if !states.is_empty() {
+            let (mut min_offset, mut max_offset) = (usize::MAX, 0);
+            for state in states {
+                let uuid = Uuid::from_slice(state.request_ctx.as_slice()).unwrap();
+                if let Some(offset) = self.contexts.remove(&uuid) {
+                    self.reads[offset].read_index = Some(state.index);
+                    min_offset = cmp::min(min_offset, offset);
+                    max_offset = cmp::max(max_offset, offset);
+                    continue;
                 }
+                error!(
+                    "cannot find corresponding read from pending reads";
+                    "uuid" => ?uuid, "read-index" => state.index,
+                );
             }
-            if self.ready_cnt < i + 1 {
-                self.ready_cnt = i + 1;
+            if max_offset > 0 {
+                self.fold(min_offset, max_offset);
+            }
+        }
+    }
+
+    fn fold(&mut self, min_offset: usize, max_offset: usize) {
+        let mut r_offset = max_offset;
+        loop {
+            let r_idx = self.reads[r_offset].read_index.unwrap();
+            for l_offset in (0..r_offset).rev() {
+                let l_idx = self.reads[l_offset].read_index.get_or_insert(r_idx);
+                match (*l_idx).cmp(&r_idx) {
+                    cmp::Ordering::Greater => *l_idx = r_idx,
+                    cmp::Ordering::Equal => continue,
```

Comment:
Should check if it's greater than min_offset before continue.

Reply:
It shouldn't be checked here, it should be checked 5 lines under.